lockfile generation still is buggy with 15.8.1

### Current Behavior

When deploying to app engine

```
╔════════════════════════════════════════════════════════════╗
╠═ Uploading 3921 files to Google Cloud Storage             ═╣
╚════════════════════════════════════════════════════════════╝
File upload done.
Updating service [default]...
....................................................................................................................................................................................failed.
ERROR: (gcloud.app.deploy) Error Response: [9] Cloud build 5331e4df-a3a8-4e40-8954-5e82f6979efe status: FAILURE
npm ERR! code EUSAGE
npm ERR! 
npm ERR! `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
npm ERR! 
npm ERR! Missing: ms@2.0.0 from lock file
```


### Expected Behavior

No error

### GitHub Repo

https://github.com/vicb/flyxc

### Steps to Reproduce

1. nx deploy - see package.json for details: `nx build fxc-server --prod --skip-nx-cache && rm dist/apps/fxc-server/package-lock.json && gcloud app deploy dist/apps/fxc-server/app.yaml`

I have to delete `dist/apps/fxc-server/package-lock.json` otherwise google cloud complains it's invalid


### Nx Report

```shell
$ nx report
>  NX  Falling back to ts-node for local typescript execution. This may be a little slower.
  - To fix this, ensure @swc-node/register and @swc/core have been installed

 >  NX   Report complete - copy this into the issue template

   Node : 18.10.0
   OS   : linux x64
   npm  : 8.19.2
   
   nx                      : 15.8.1
   @nrwl/js                : 15.8.1
   @nrwl/jest              : 15.8.1
   @nrwl/linter            : 15.8.1
   @nrwl/workspace         : 15.8.1
   @nrwl/cli               : 15.8.1
   @nrwl/cypress           : 15.8.1
   @nrwl/devkit            : 15.8.1
   @nrwl/eslint-plugin-nx  : 15.8.1
   @nrwl/express           : 15.8.1
   @nrwl/node              : 15.8.1
   @nrwl/rollup            : 15.8.1
   @nrwl/tao               : 15.8.1
   @nrwl/vite              : 15.8.1
   @nrwl/web               : 15.8.1
   @nrwl/webpack           : 15.8.1
   typescript              : 4.9.5
   ---------------------------------------
   Community plugins:
   @nx-tools/nx-container : 4.0.3
```


### Failure Logs

_No response_

### Additional Information

_No response_

I'm having a similar issue with `npm ci` on Nx 15.8.2

```shell
npm ERR! Missing: long@4.0.0 from lock file
npm ERR! Missing: ms@2.0.0 from lock file
```

## NX Report
```
 >  NX   Report complete - copy this into the issue template

   Node : 18.12.1
   OS   : darwin arm64
   npm  : 8.19.2

   nx                      : 15.8.2
   @nrwl/js                : 15.8.2
   @nrwl/jest              : 15.8.2
   @nrwl/linter            : 15.8.2
   @nrwl/workspace         : 15.8.2
   @nrwl/angular           : 15.8.2
   @nrwl/cli               : 15.8.2
   @nrwl/cypress           : 15.8.2
   @nrwl/devkit            : 15.8.2
   @nrwl/eslint-plugin-nx  : 15.8.2
   @nrwl/nest              : 15.8.2
   @nrwl/node              : 15.8.2
   @nrwl/tao               : 15.8.2
   @nrwl/webpack           : 15.8.2
   @nrwl/nx-cloud          : 15.1.1
   typescript              : 4.9.5
   ---------------------------------------
   Community plugins:
   apollo-angular         : 4.2.1
   @nx-tools/nx-container : 4.0.3
```

## npm ls ms (after npm i)
```
├─┬ @nestjs/platform-express@9.0.0
│ ├─┬ body-parser@1.20.0
│ │ └─┬ debug@2.6.9
│ │   └── ms@2.0.0
│ └─┬ express@4.18.1
│   ├─┬ finalhandler@1.2.0
│   │ └─┬ debug@2.6.9
│   │   └── ms@2.0.0
│   └─┬ send@0.18.0
│     ├─┬ debug@2.6.9
│     │ └── ms@2.0.0
│     └── ms@2.1.3
├─┬ @sentry/node@7.38.0
│ └─┬ https-proxy-agent@5.0.1
│   └─┬ debug@4.3.4
│     └── ms@2.1.2 deduped
├─┬ apollo-server-express@3.11.1
│ ├─┬ body-parser@1.20.1
│ │ └─┬ debug@2.6.9
│ │   └── ms@2.0.0
│ └─┬ express@4.18.2
│   └─┬ debug@2.6.9
│     └── ms@2.0.0
└─┬ jsonwebtoken@9.0.0
  └── ms@2.1.2
```

## npm ls long (after npm i)
```
├─┬ @nestjs/apollo@10.1.7
│ └─┬ apollo-server-core@3.11.1
│   ├─┬ @apollo/utils.usagereporting@1.0.1
│   │ └─┬ @apollo/usage-reporting-protobuf@4.0.2
│   │   └─┬ @apollo/protobufjs@1.2.7
│   │     └── long@4.0.0 deduped
│   └─┬ apollo-reporting-protobuf@3.3.3
│     └─┬ @apollo/protobufjs@1.2.6
│       └── long@4.0.0 deduped
└─┬ @nestjs/terminus@9.2.1
  ├─┬ @grpc/grpc-js@1.7.3
  │ └─┬ @grpc/proto-loader@0.7.4
  │   ├── long@4.0.0
  │   └─┬ protobufjs@7.2.0
  │     └── long@5.2.1
  └─┬ @grpc/proto-loader@0.6.13
    ├── long@4.0.0
    └─┬ protobufjs@6.11.3
      └── long@4.0.0 deduped
```

After `npm i` at lot of `optional: true` were added to the `package-lock.json` after integrity sha.

Same issue on `15.8.3`, with error:

```code EUSAGE
npm ERR! 
npm ERR! `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
npm ERR! 
npm ERR! Missing: ms@2.0.0 from lock file

--- a/packages/nx/src/lock-file/npm-parser.ts
+++ b/packages/nx/src/lock-file/npm-parser.ts
@@ -413,7 +413,7 @@ function mapSnapshots(
   const visitedNodes = new Map<ProjectGraphExternalNode, Set<string>>();
   const visitedPaths = new Set<string>();
 
-  const remappedPackages: MappedPackage[] = [];
+  const remappedPackages: Map<string, MappedPackage> = new Map();
 
   // add first level children
   Object.values(graph.externalNodes).forEach((node) => {
@@ -423,7 +423,7 @@ function mapSnapshots(
         node.data.packageName,
         node.data.version
       );
-      remappedPackages.push(mappedPackage);
+      remappedPackages.set(mappedPackage.path, mappedPackage);
       visitedNodes.set(node, new Set([mappedPackage.path]));
       visitedPaths.add(mappedPackage.path);
     } else {
@@ -431,6 +431,7 @@ function mapSnapshots(
     }
   });
 
+  let remappedPackagesArray: MappedPackage[];
   if (nestedNodes.size) {
     const invertedGraph = reverse(graph);
     nestMappedPackages(
@@ -441,9 +442,13 @@ function mapSnapshots(
       visitedPaths,
       rootLockFile
     );
+    // initially we naively map package paths to topParent/../parent/child
+    // but some of those should be nested higher up the tree
+    remappedPackagesArray = elevateNestedPaths(remappedPackages);
+  } else {
+    remappedPackagesArray = Array.from(remappedPackages.values());
   }
-
-  return remappedPackages.sort((a, b) => a.path.localeCompare(b.path));
+  return remappedPackagesArray.sort((a, b) => a.path.localeCompare(b.path));
 }
 
 function mapPackage(
@@ -452,8 +457,8 @@ function mapPackage(
   version: string,
   parentPath = ''
 ): MappedPackage {
-  const path = parentPath + `node_modules/${packageName}`;
   const lockfileVersion = rootLockFile.lockfileVersion;
+
   let valueV3, valueV1;
   if (lockfileVersion < 3) {
     valueV1 = findMatchingPackageV1(
@@ -471,7 +476,7 @@ function mapPackage(
   }
 
   return {
-    path,
+    path: parentPath + `node_modules/${packageName}`,
     name: packageName,
     valueV1,
     valueV3,
@@ -480,7 +485,7 @@ function mapPackage(
 
 function nestMappedPackages(
   invertedGraph: ProjectGraph,
-  result: MappedPackage[],
+  result: Map<string, MappedPackage>,
   nestedNodes: Set<ProjectGraphExternalNode>,
   visitedNodes: Map<ProjectGraphExternalNode, Set<string>>,
   visitedPaths: Set<string>,
@@ -499,15 +504,13 @@ function nestMappedPackages(
 
       if (visitedNodes.has(targetNode)) {
         visitedNodes.get(targetNode).forEach((path) => {
-          const parentPath =
-            findParentPath(path, node.data.packageName, visitedPaths) + '/';
           const mappedPackage = mapPackage(
             rootLockFile,
             node.data.packageName,
             node.data.version,
-            parentPath
+            path + '/'
           );
-          result.push(mappedPackage);
+          result.set(mappedPackage.path, mappedPackage);
           if (visitedNodes.has(node)) {
             visitedNodes.get(node).add(mappedPackage.path);
           } else {
@@ -537,21 +540,59 @@ function nestMappedPackages(
   }
 }
 
-function findParentPath(
-  path: string,
-  packageName: string,
-  visitedPaths: Set<string>
-): string {
-  const segments = path.split('/node_modules/');
-  let parentPath = path;
-  while (
-    segments.length > 1 &&
-    !visitedPaths.has(`${parentPath}/node_modules/${packageName}`)
-  ) {
-    segments.pop();
-    parentPath = segments.join('/node_modules/');
-  }
-  return parentPath;
+// sort paths by number of segments and then alphabetically
+function sortMappedPackagesPaths(mappedPackages: Map<string, MappedPackage>) {
+  return Array.from(mappedPackages.keys()).sort((a, b) => {
+    const aLength = a.split('/node_modules/').length;
+    const bLength = b.split('/node_modules/').length;
+    if (aLength > bLength) {
+      return 1;
+    }
+    if (aLength < bLength) {
+      return -1;
+    }
+    return a.localeCompare(b);
+  });
+}
+
+function elevateNestedPaths(
+  remappedPackages: Map<string, MappedPackage>
+): MappedPackage[] {
+  const result = new Map<string, MappedPackage>();
+  const sortedPaths = sortMappedPackagesPaths(remappedPackages);
+
+  sortedPaths.forEach((path) => {
+    const segments = path.split('/node_modules/');
+
+    // we keep hoisted packages intact
+    if (segments.length === 1) {
+      result.set(path, remappedPackages.get(path));
+      return;
+    }
+
+    const packageName = segments.pop();
+    const getNewPath = (segs) =>
+      `${segs.join('/node_modules/')}/node_modules/${packageName}`;
+
+    // check if grandparent has the same package
+    while (
+      segments.length > 1 &&
+      !result.has(getNewPath(segments.slice(0, -1)))
+    ) {
+      segments.pop();
+    }
+    const newPath = getNewPath(segments);
+    if (path !== newPath) {
+      result.set(newPath, {
+        ...remappedPackages.get(path),
+        path: newPath,
+      });
+    } else {
+      result.set(path, remappedPackages.get(path));
+    }
+  });
+
+  return Array.from(result.values());
 }
 
 function findMatchingPackageV3(

